plugin/evm/header/extra.go

```diff
@@ -66,14 +157,49 @@ func VerifyExtra(rules extras.AvalancheRules, extra []byte) error {
 			)
 		}
 	default:
-		if uint64(extraLen) > params.MaximumExtraDataSize {
+		if uint64(extraLen) > MaximumExtraDataSize {
 			return fmt.Errorf(
 				"%w: expected <= %d but got %d",
 				errInvalidExtraLength,
-				params.MaximumExtraDataSize,
+				MaximumExtraDataSize,
 				extraLen,
 			)
 		}
 	}
 	return nil
 }
+
+// PredicateBytesFromExtra returns the predicate result bytes from the header's
+// extra data. If the extra data is not long enough, an empty slice is returned.
+func PredicateBytesFromExtra(rules extras.AvalancheRules, extra []byte) []byte {
+	offset := ap3.WindowSize
+	if rules.IsFortuna {
+		offset = acp176.StateSize
+	}
+
+	// Prior to Durango, the VM enforces the extra data is smaller than or equal
+	// to `offset`.
+	// After Durango, the VM pre-verifies the extra data past `offset` is valid.
+	if len(extra) <= offset {
+		return nil
+	}
+	return extra[offset:]
+}
+
+// SetPredicateBytesInExtra sets the predicate result bytes in the header's extra
+// data. If the extra data is not long enough (i.e., an incomplete header.Extra
+// as built in the miner), it is padded with zeros.
+func SetPredicateBytesInExtra(rules extras.AvalancheRules, extra []byte, predicateBytes []byte) []byte {
+	offset := ap3.WindowSize
+	if rules.IsFortuna {
+		offset = acp176.StateSize
+	}
+
+	if len(extra) < offset {
+		// pad extra with zeros
+		extra = append(extra, make([]byte, offset-len(extra))...)
+	}
+	extra = extra[:offset] // truncate extra to the offset
```

Comment:
I wonder if we should just truncate if len(extra) > offset here (in else above)

Reply:
It's the same thing really. I like the else approach you mention a bit better for clarity, but, given how rebasing this branch is mega troublesome (and we want to merge --ff-only it in libevm), I'll skip this unfortunately!

Reply:
I have it in a local stash, I'll try applying a bunch of small changes to the libevm branch later!

Reply:
I'm again confused with what you mean by rebasing. I'd like to keep those changes in the same PR if it's changed in the same PR. Alternatively we can have review PRs but IMHO they should be opened before we close base PRs.